### ISSUE TITLE
resolved should accept DNS responses with incorrect RR count

### ISSUE THREAD

**Issue by moocan** [systemd/systemd#12841]
**systemd version the issue has been seen with**
> systemd 234
> systemd 237

**Used distribution**
 > Ubuntu 17.10 => systemd 234
> Ubuntu LTS 18.04 => systemd 237

**Expected behaviour you didn't see**
 > systemd-resolve -i enp1s0 BRN3C2AF403CBF0.mydomain.ext
BRN3C2AF403CBF0.mydomain.ext: 10.0.0.2
-- Information acquired via protocol DNS in 5.1ms.
-- Data is authenticated: no

It is working sometimes about 1/1000 and when it is working it is working for 30 mins max and the stop. Restarting systemd-resolved or unplugging cable has no effect. 
When it is working (SYSTEMD_LOG_LEVEL=debug):
```
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_call sender=:1.1153 destination=org.freedesktop.resolve1 object=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Looking up RR for BRN3C2AF403CBF0.mydomain.ext IN A.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Looking up RR for BRN3C2AF403CBF0.mydomain.ext IN AAAA.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=63 reply_cookie=0 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=48 reply_cookie=63 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetNameOwner cookie=64 reply_cookie=0 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=49 reply_cookie=64 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Cache miss for BRN3C2AF403CBF0.mydomain.ext IN A
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Transaction 27736 for <BRN3C2AF403CBF0.mydomain.ext IN A> scope dns on enp1s0/*.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Using feature level UDP for transaction 27736.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Using DNS server 10.0.0.254 for transaction 27736.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Sending query packet with id 27736.
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Processing incoming packet on transaction 27736. (rcode=SUCCESS)
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Not caching zero TTL cache entry: BRN3C2AF403CBF0.mydomain.ext IN A
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Transaction 27736 for <BRN3C2AF403CBF0.mydomain.ext IN A> on scope dns on enp1s0/* now complete with <success> from network (unsigned).
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_return sender=n/a destination=:1.1153 object=n/a interface=n/a member=n/a cookie=65 reply_cookie=2 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=66 reply_cookie=0 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=50 reply_cookie=66 error=n/a
juin 20 06:03:16 eita.mydomain.ext systemd-resolved[13680]: Freeing transaction 27736.
```

**Unexpected behaviour you saw**
 > systemd-resolve -i enp1s0 BRN3C2AF403CBF0.mydomain.ext
BRN3C2AF403CBF0.mydomain.ext: resolve call failed: Received invalid reply

When it is not working (SYSTEMD_LOG_LEVEL=debug):

```
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_call sender=:1.1137 destination=org.freedesktop.resolve1 object=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Looking up RR for eita.mydomain.ext IN A.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Looking up RR for eita.mydomain.ext IN AAAA.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=8 reply_cookie=0 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=9 reply_cookie=8 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetNameOwner cookie=9 reply_cookie=0 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=10 reply_cookie=9 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Cache miss for eita.mydomain.ext IN A
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Transaction 2742 for <eita.mydomain.ext IN A> scope dns on enp1s0/*.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Using feature level UDP+EDNS0 for transaction 2742.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Using DNS server 10.0.0.254 for transaction 2742.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Sending query packet with id 2742.
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Processing incoming packet on transaction 2742. (rcode=SUCCESS)
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Transaction 2742 for <eita.mydomain.ext IN A> on scope dns on enp1s0/* now complete with <invalid-reply> from network (unsigned).
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Sent message type=error sender=n/a destination=:1.1137 object=n/a interface=n/a member=n/a cookie=10 reply_cookie=2 error=Received invalid reply
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=11 reply_cookie=0 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1135 object=n/a interface=n/a member=n/a cookie=11 reply_cookie=11 error=n/a
juin 20 05:42:52 eita.mydomain.ext systemd-resolved[13680]: Freeing transaction 2742.
```
/etc/resolv.conf
```
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# 127.0.0.53 is the systemd-resolved stub resolver.
# run "systemd-resolve --status" to see details about the actual nameservers.
nameserver 127.0.0.53

search mydomain.ext
```
systemd-resolve --status:
```
Global
          DNS Domain: mydomain.ext
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa
                      18.172.in-addr.arpa
                      19.172.in-addr.arpa
                      20.172.in-addr.arpa
                      21.172.in-addr.arpa
                      22.172.in-addr.arpa
                      23.172.in-addr.arpa
                      24.172.in-addr.arpa
                      25.172.in-addr.arpa
                      26.172.in-addr.arpa
                      27.172.in-addr.arpa
                      28.172.in-addr.arpa
                      29.172.in-addr.arpa
                      30.172.in-addr.arpa
                      31.172.in-addr.arpa
                      corp
                      d.f.ip6.arpa
                      home
                      internal
                      intranet
                      lan
                      local
                      private
                      test

Link 3 (wlp2s0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 193.183.98.66
                      51.254.25.115
                      188.165.200.156

Link 2 (enp1s0)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 10.0.0.254
          DNS Domain: mydomain.ext

```

**Steps to reproduce the problem**

The DNS Server is an embedded system from an ISP provider.
Working: `Using feature level UDP for transaction ...`
Not Working: `Using feature level UDP+EDNS0 for transaction ...`

`dig NPIC92ABD.mydomain.ext @10.0.0.254` is always returning a result.

BUT using tcpdump / wireshark
`dig NPIC92ABD.mydomain.ext @10.0.0.254` revealed a [Malformed packet] in the capture but is returning the result while all is ok with `dig +noedns NPIC92ABD.mydomain.ext @10.0.0.254` 

Is it possible using NetworkManager or systemd-resolved settings to force systemd-resolved to use `UDP` only and not `UDP+EDNS0` _only for this DNS Server_ which cause systemd-resolved to report `invalid-reply` while dig is always returning an answer ?
 
Kind Regards

**Comment by james-choncholas** [comment]
I have observed the same issue and am also looking for a way to force systemd-resolved to use UDP only and not UDP+EDNS0 for a specific DNS server

**Comment by skwlilac** [comment]
@james-choncholas I have the same issue. Did you find away to disable the +EDNS0 either per interface or globally. Thanks.

**Comment by poettering** [comment]
Is this reproducible on current systemd versions? If so, it would be great to have a pcap file for one of these transactions, to figure out why resolved finds the reply invalid.

**Comment by skwlilac** [comment]
@poettering AFAICT it is still not working for me (an upto date Ubunto 20.04). For me the problem arises when connected via VPN to our premises router. Attempts to resolve a name for a local dect box yield the following lines in syslog:

```
ov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Got DNS stub UDP query packet for id 19568
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Looking up RR for dect.internal.epimorphics.net IN A.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Cache miss for dect.internal.epimorphics.net IN A
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Transaction 38630 for <dect.internal.epimorphics.net IN A> scope dns on tun0/*.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Using feature level UDP+EDNS0+DO+LARGE for transaction 38630.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Using DNS server 10.105.1.1 for transaction 38630.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Sending query packet with id 38630.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Processing query...
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Processing incoming packet on transaction 38630 (rcode=SUCCESS).
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Transaction 38630 for <dect.internal.epimorphics.net IN A> on scope dns on tun0/* now complete with <invalid-reply> from network (unsigned).
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Sending response packet with id 19568 on interface 1/AF_INET.
Nov  9 14:44:09 skw-Ubuntu-VBox systemd-resolved[5624]: Freeing transaction 38630.
```

I've attached a .pcap trace from the tunnel packets 2043 and 2046 are probably the pair that you want. The rest seem to be a blizzard of IPv6 inverse address lookup (not sure what's causing that) the underlying local network supports IPv6, but neither the tunnel or the remote end of the VPN connection do. The remote end of the VPN connection is an aging cisco small business router (RV3xx series).

[resolved.zip](https://github.com/systemd/systemd/files/5511300/resolved.zip)

Hope that's of some help.

**Comment by poettering** [comment]
Hmm , so my wireshark also considers the response packets invalid. it adds a big marker [Malformed Packet] on it. Doesn't it when you look at this?

The response claims to contain 2 response RRs, but only contains one. And TC isn't set either.

Looking at this, resolved, wireshark and I myself all agree the packet is invalid. Smells like a bug in the relevant server?

If this is a common issue that other DNS clients are OK with we could probably tweak our parser to handle this better. But if it's an exotic one-off issue, it's probably better fixed in the offending DNS server?

**Comment by skwlilac** [comment]
For me the problem is that the request is made with EDNS0 set to a server (our premise router) which is of a vintage that doesn't support EDNS0. My personal workaround for the last several months has been to not use systemd-resolved and instead rerun with dnsmasq. Doing some more googling yesterday I found this:

https://bugs.launchpad.net/ubuntu/+source/resolvconf/+bug/1817903

which seems to echo exactly the problem I'm experiencing. Whilst the thread there seems to indicate that things got fixed around  April 2019, and that may be true for fresh installations, but mine has been serially upgraded from about 16.04. Also I'm no guru with system configuration in terms and it seems that there's potentially some interaction between resolvconf, systemd-networkd and NetworkManager that may need addressing - spending time fiddling with the config isn't a priority for me right now as I have a way of working that works for me.

Anyway to cut to the chase, yes I think you are right the response packets are invalid and AIUI that's because the request was made with an EDSN0 option that the device does not support.

**Comment by poettering** [comment]
@moocan any chance you can provide me with the pcap data of your case?

**Comment by moocan** [comment]
[traces.zip](https://github.com/systemd/systemd/files/5541543/traces.zip)
Sorry for the delay.
Here what I can provide to you quickly.
You will find in attachment two network traces.
- dig+noedns-working case
- dig-working case containing Malformed packet and the issue with systemd-resolved

If you need something else I remain fully available.
I had two major issues with systemd-resolved (this one and the #15202) and found any workaround any kind of control over it so I had no choice but disable system-resolved from all computers and use dnsmasq.

Kind Regards

**Comment by poettering** [comment]
Trace shows the same issue: RR count in response is simply wrong. #17535 contains a work-around for handling htat.
